### PR TITLE
fix(core): fix UTF-8 handling issues in logging

### PR DESCRIPTION
**Pull request by amunra**
Fix the following UTF-8 related bugs in our logging infra:
* When logging with the default (non sync) `Logger`, the `$utf8(long lo, long hi)` method now correctly copies the UTF-8 bytes to the logger, rather than first decoding to UTF-16 and then discarding the high byte and displaying incorrect text. This has now been further optimized to almost always just invoke `Unsafe.copyMemory`.
* Fixed `LogRecordSink` logic to truncate log lines at code point rather than byte boundaries: This avoids generating log files that can't be read as they would contain invalid UTF-8.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 26 / 30 (86.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/log/LogRecordSink.java|24|28|85.71%|
|:large_blue_circle:|io/questdb/log/Logger.java|2|2|100.00%|

**Comment by amunra** [comment]
I'm pretty sure that in `LogRecordSink.java`

```java
    @Override
    public CharSink put(CharSequence cs, int lo, int hi) {
        int rem = (int) (lim - _wptr);
        int len = hi - lo;
        int n = Math.min(rem, len);
        Chars.asciiStrCpy(cs, lo, n, _wptr);
        _wptr += n;
        return this;
    }
```

is also broken and can log a broken UTF-8 sequence, but I honestly ran out of steam.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 28 / 28 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/log/LogRecordSink.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|2|2|100.00%|

File: core/src/main/java/io/questdb/log/LogRecordSink.java [hunk 1]
```diff
@@ -79,8 +82,25 @@ public CharSink put(CharSequence cs, int lo, int hi) {
 
     @Override
     public CharSink put(char c) {
-        if (_wptr < lim) {
-            Unsafe.getUnsafe().putByte(_wptr++, (byte) c);
+        if (done) {
+            return this;
+        }
+
+        final long left = lim - _wptr;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/log/LogRecordSink.java:89]: we need fast path optimisation here, 99.9999999% of what we log is ascii

File: core/src/main/java/io/questdb/log/LogRecordSink.java [hunk 2]
```diff
@@ -79,8 +82,25 @@ public CharSink put(CharSequence cs, int lo, int hi) {
...
+        if (needed < 1) {
+            // Invalid UTF-8 byte, sentinel replacement.
+            b = (byte) '?';
+            needed = 1;
+        }
+
+        if (left >= needed) {
+            Unsafe.getUnsafe().putByte(_wptr++, b);
+        }
+        else {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/log/LogRecordSink.java:101]: code formatting

File: core/src/main/java/io/questdb/log/LogRecordSink.java [hunk 3]
```diff
@@ -79,8 +82,25 @@ public CharSink put(CharSequence cs, int lo, int hi) {
...
+            // Stop processing any further characters.
+            done = true;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/log/LogRecordSink.java:103]: why is this needed? comments are necessart

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by bluestreak01** [comment]
PR title should be in present tense and not capitalized 

image

**Comment by bluestreak01** [comment]
@amunra PR description is required

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 85 / 88 (96.59%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/str/CharSink.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/log/LogRecordSink.java|84|85|98.82%|
|:large_blue_circle:|io/questdb/log/Logger.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 87 / 88 (98.86%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/log/LogRecordSink.java|84|85|98.82%|
|:large_blue_circle:|io/questdb/std/str/CharSink.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 88 / 88 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/str/CharSink.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/log/LogRecordSink.java|85|85|100.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|1|1|100.00%|

File: core/src/main/java/io/questdb/log/LogRecordSink.java [hunk 4]
```diff
@@ -59,28 +63,121 @@ public int length() {
...
+            Chars.asciiStrCpy(cs, len, _wptr);
+            _wptr += len;
+            return this;
+        }
+
+        // The line is being truncated:
+        // We determine a safe length to byte-copy.
+        // We skip copying the last 4 bytes, as they may be a multi-byte UTF-8 character.
+        // NOTE: The computed length may be negative.
+        int safeLen = rem - 4;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/log/LogRecordSink.java:79]: I guess there was confusion as to what this method is used for.

Context: this is logger and not a generic one, logger for QuestDB. All our fixed strings are English ASCII. E.g. There are no deliberate strings constants that are non-ascii.

There are two cases where we may have anything non-ascii:
1. UTF16 encoded CharSequence, in which case code must explicitly call `log.utf8(CharSequence sequence)`
2. Somethings on the wire in UTF8 format, which we would like to log correctly. In this case we call your function `log.utf8(lo,hi)`

Therefore assumptions about multi-byte characters here will not hold true

File: core/src/main/java/io/questdb/log/LogRecordSink.java [hunk 5]
```diff
@@ -59,28 +63,121 @@ public int length() {
...
     }
 
     @Override
     public CharSink put(CharSequence cs, int lo, int hi) {
-        int rem = (int) (lim - _wptr);
-        int len = hi - lo;
-        int n = Math.min(rem, len);
-        Chars.asciiStrCpy(cs, lo, n, _wptr);
-        _wptr += n;
+        final int rem = (int) (lim - _wptr);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/log/LogRecordSink.java:96]: same, this is ascii-only function

File: core/src/main/java/io/questdb/log/LogRecordSink.java [hunk 6]
```diff
@@ -59,28 +63,121 @@ public int length() {
...
+            put(cs.charAt(lo + i));
+        }
         return this;
     }
 
     @Override
     public CharSink put(char c) {
-        if (_wptr < lim) {
-            Unsafe.getUnsafe().putByte(_wptr++, (byte) c);
+        final long left = lim - _wptr;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/log/LogRecordSink.java:119]: i don't believe this function is used for anything other than bytes, was it broken? What motivated the change?

To illustrate what i am saying if one calls `log.utf8(CharSequence)` single char that is 4 bytes log, the `utf8()` will call `put(char)` 4 times.

File: core/src/test/java/io/questdb/log/LogRecordSinkTest.java [hunk 7]
```diff
@@ -122,4 +251,163 @@ public void testSimpleMessage3() throws Exception {
...
+    /**
+     * A class that holds UTF-8 in native memory.
+     * <p>
+     * Solely intended for testing purposes.
+     * <p>
+     * It conforms to the expections of LogRecordSink
+     * and its length() / charAt() etc methods operate/return
+     * bytes even when the signature says chars.
+     */
+    static final class Utf8 implements LPSZ, CharSequence, Closeable {
```
Comment (bluestreak01) [inline core/src/test/java/io/questdb/log/LogRecordSinkTest.java:343]: i don't believe we have anything that behaves like this class, pls correct me if i'm wrong

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/log/LogRecordSink.java (hunk 4 above)
Comment (amunra) [reply core/src/main/java/io/questdb/log/LogRecordSink.java]: Rolled back changes for `put(CharSequence, ..)` after your clarification on their use case (and some digging to confirm it).

File: core/src/main/java/io/questdb/log/LogRecordSink.java (hunk 6 above)
Comment (amunra) [reply core/src/main/java/io/questdb/log/LogRecordSink.java:105]: The function is indeed only ever used for bytes.. but not all these bytes are 7-bit ASCII. Sometimes these are UTF-8 bytes (cast to `char`).
The validation here is needed.

File: core/src/main/java/io/questdb/log/LogRecordSink.java (hunk 5 above)
Comment (amunra) [reply core/src/main/java/io/questdb/log/LogRecordSink.java]: Yes, rolled back.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 67 / 67 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/str/CharSink.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/log/LogRecordSink.java|64|64|100.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|1|1|100.00%|

File: core/src/main/java/io/questdb/log/LogRecordSink.java [hunk 8]
```diff
@@ -57,8 +61,19 @@ public int length() {
...
+    /**
+     * Log a string that contains only ASCII chars.
+     * Call `encodeUtf8` otherwise.
+     */
     @Override
     public CharSink put(CharSequence cs) {
+        for (int i = 0, n = cs.length(); i < n; i++) {
+            char c = cs.charAt(i);
+            boolean isAscii = (c >= 0) && (c <= 127);
+            if (!isAscii) {
```
[the remark this answers is not in the document]
Comment (amunra) [reply core/src/main/java/io/questdb/log/LogRecordSink.java]: Done.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 62 / 62 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/str/CharSink.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/log/LogRecordSink.java|59|59|100.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 62 / 62 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/str/CharSink.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/log/LogRecordSink.java|59|59|100.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|1|1|100.00%|

**Review by bluestreak01 — DISMISSED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 62 / 62 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/str/CharSink.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/log/LogRecordSink.java|59|59|100.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|1|1|100.00%|

**Review by bluestreak01 — APPROVED** [review]
